### PR TITLE
build(deps): upgrade opendal to 0.46

### PR DESCRIPTION
**Pull request by tisonkun**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

STILL SOME BREAKING CHANGES.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/object-store/Cargo.toml
Comment (tisonkun) [inline src/object-store/Cargo.toml:30]: The default features is trimeed as `"services-memory", "reqwest/rustls"` that is exactly we want.

**Comment by Xuanwo** [comment]
Hi, in https://github.com/tisonkun/greptimedb/pull/5, I make greptime buildable with opendal 0.46.

We still have more work to do moving forward. For example, we need to reconsider the use of `AsyncRead` + `AsyncSeek` in our codebase, as they are not suitable for object storage services and upcoming completion-based IO.

Also, since opendal 0.46, it returns [`Buffer`](https://opendal.apache.org/docs/rust/opendal/struct.Buffer.html) which provide zero-copy of data. We can try use this feature instead of calling `to_vec`, `to_bytes` everywhere. For example, our decoder can accept `std::io::Read` instead of `&[u8]`.

There are some places for opendal to improve like `into_futures_async_read` requires `Range<u64>`. We have fixed it and will be included in next release.

**Comment by Xuanwo** [comment]
For PrometheusLayer, I raised https://github.com/GreptimeTeam/greptimedb/issues/4041

**Comment by tisonkun** [comment]
@Xuanwo seems we have two test cases failed:

* test_orc_opener - Footer length is empty
* test_object_store_cache_policy - file not found: ecfe0dce85de452eb0a325158e7bfb75.cache-bytes=7-4194310

The former seems related to ORC read content, the latter seems related to range encoding.

Would you help take a look?

**Comment by Xuanwo** [comment]
cache test should be fixed by https://github.com/tisonkun/greptimedb/pull/6

**Comment by tisonkun** [comment]
@Xuanwo Thank you and integrated. Could you cross-link the OpenDAL PR that can fix the hack you mention in https://github.com/tisonkun/greptimedb/pull/6?

**Comment by Xuanwo** [comment]
> @Xuanwo Thank you and integrated. Could you cross-link the OpenDAL PR that can fix the hack you mention in [tisonkun#6](https://github.com/tisonkun/greptimedb/pull/6)?

It happens in https://github.com/apache/opendal/pull/4634

File: src/object-store/tests/object_store_test.rs
Comment (tisonkun) [inline src/object-store/tests/object_store_test.rs:1]: @Xuanwo why the result of these cases changed?

Comment (Xuanwo) [reply src/object-store/tests/object_store_test.rs:1]: > @Xuanwo why the result of these cases changed?

As mentioned in https://github.com/GreptimeTeam/greptimedb/pull/4037#issuecomment-2131108621

OpenDAL will never send `0..` anymore.

**Comment by killme2008** [comment]
It seems that there are a few tasks waiting for the opendal 0.4.7 release. I believe we should postpone this PR until then.

**Comment by tisonkun** [comment]
> It seems that there are a few tasks waiting for the opendal 0.4.7 release. I believe we should postpone this PR until then.

@Xuanwo Is this regression from 0.45 to 0.46?

* https://github.com/GreptimeTeam/greptimedb/pull/4037#issuecomment-2131111347

If not, I'd prefer we upgrade to 0.46, and then 0.47 as it's expected to release in the next few weeks:

* https://lists.apache.org/thread/s4w9okds2mknhjgkdcxyhy2nbyr0yd6d
* https://github.com/apache/opendal/issues/4640

**Comment by Xuanwo** [comment]
> Is this regression from 0.45 to 0.46?

No, I don't think it's a regression. The current behavior is expected. OpenDAL 0.47 simply introduces some additional improvements, for example, allow users calling `into_futures_async_read` with size.

**Review by evenyag — APPROVED** [review]
@Xuanwo Thanks for helping us to upgrade OpenDAL. I left some non-blocking comments to remind us after we merge this PR.

File: src/mito2/src/cache/write_cache.rs
Comment (Xuanwo) [reply src/mito2/src/cache/write_cache.rs:202]: Yep!

File: src/datanode/src/store.rs
Comment (tisonkun) [reply src/datanode/src/store.rs:110]: Addressed.

File: src/datanode/src/store.rs
Comment (tisonkun) [reply src/datanode/src/store.rs:140]: Somehow related to the hyper-1.0 upgrade. I'd take a look or postpone it as a follow-up.

Comment (Xuanwo) [reply src/datanode/src/store.rs:140]: This could be a massive project; I spent an entire workweek on it.

image

File: src/mito2/src/cache/write_cache.rs
Comment (tisonkun) [reply src/mito2/src/cache/write_cache.rs:202]: @Xuanwo is it available in 0.46? If so, I'd appreciate it if you can send a patch to this PR >_<

File: src/mito2/src/manifest/storage.rs
Comment (tisonkun) [reply src/mito2/src/manifest/storage.rs:429]: I found the `checkpoint` itself is a `Vec<u8>`. Let me try to remove the `to_vec()` here.

File: src/mito2/src/cache/write_cache.rs
Comment (Xuanwo) [reply src/mito2/src/cache/write_cache.rs:202]: > is it available in 0.46?

`into_bytes_stream` & `into_bytes_sink` is available in v0.46. `into_stream` & `into_sink` will be available in v0.47.

> If so, I'd appreciate it if you can send a patch to this PR >_<

I'm willing to start a new PR instead of a patch to this PR. This PR is already too large.

File: src/mito2/src/manifest/storage.rs
Comment (tisonkun) [reply src/mito2/src/manifest/storage.rs:429]: No. It's an IDE fault that resolved to 0.45. The new type is `Buffer`.

File: src/mito2/src/cache/write_cache.rs
Comment (tisonkun) [reply src/mito2/src/cache/write_cache.rs:202]: Make sense.

**Comment by tisonkun** [comment]
@killme2008 as this won't introduce explicit regression. I'd  prefer we upgrade to 0.46, and then 0.47 as it's expected to release in the next few weeks.

See https://github.com/GreptimeTeam/greptimedb/pull/4037#issuecomment-2132514648.

File: src/mito2/src/manifest/storage.rs
Comment (tisonkun) [reply src/mito2/src/manifest/storage.rs]: Finished at 90bbb0bf2231173bee130c32c50f63a4483cb856.

**Comment by WenyXu** [comment]
By the way, I'm fixing the failed CI in #4049.🥲

File: src/datanode/src/store.rs
Comment (WenyXu) [inline src/datanode/src/store.rs:170]: cc code owner @killme2008

Comment (tisonkun) [reply src/datanode/src/store.rs:170]: This seems a regression. cc @Xuanwo 

I'll take a look also.

Comment (Xuanwo) [reply src/datanode/src/store.rs:170]: I left a `FIXME` here, it's not a regression.

Comment (tisonkun) [reply src/datanode/src/store.rs:170]: Let me try to upgrade it or use `reqwest_012 = { package = "reqwest", version = "0.12" }` here.

Comment (tisonkun) [reply src/datanode/src/store.rs]: Resolved at https://github.com/GreptimeTeam/greptimedb/pull/4037/commits/ad0793b48cb0ac19fbd7c52be215140d96ca6eb5.

File: src/datanode/src/store.rs
Comment (tisonkun) [reply src/datanode/src/store.rs]: Somehow make it work at https://github.com/GreptimeTeam/greptimedb/pull/4037/commits/ad0793b48cb0ac19fbd7c52be215140d96ca6eb5.

We still need to bump axum, tonic, hyper, http and more. But bridge reqwest seems separated.

File: src/mito2/src/sst/index/applier.rs
Comment (tisonkun) [reply src/mito2/src/sst/index/applier.rs:1]: Let's make a separated issue for this?

Comment (evenyag) [reply src/mito2/src/sst/index/applier.rs:1]: Maybe a discussion?

Comment (tisonkun) [reply src/mito2/src/sst/index/applier.rs:1]: Make sense.

**Review by WenyXu — APPROVED** [review]
LGTM

**Pull request merged**
